design-documents/asynchronous-operations/consumer-processes-improvements.md

```diff
@@ -0,0 +1,86 @@
+
+Currently all defined message queue consumers are getting spawned by a cronjob called `consumers_runner`. You can optionally choose to disable this entirely or only allow specific consumers to run based on some config settings in the `app/etc/env.php` file (see [docs](https://devdocs.magento.com/guides/v2.3/config-guide/mq/manage-message-queues.html#configuration)).  
+This `consumers_runner` job first looks around on the server(s) to see if there are already running consumer processes for the same queue. If that's the case it won't spawn a new one, but if no consumer process is found for a certain queue, it will spawn one. By default it will listen for maximum 10.000 messages and if those 10.000 messages have been handled, it will kill itself. The next time the cronjob `consumers_runner` triggers, it will spawn a new process.
+
+There is also a poison pill feature in Magento, which is basically just a random hash stored in the database in the table `queue_poison_pill`. If some configuration value gets changed using the Magento backend, the poison pill version is changed. 
+The running consumer processes check on this version, but only when a new message is in the queueu. If a consumer discovers a new message in the queue, it first checks if the poison pill version has changed. If the version wasn't changed, the message will be handled. If the version did get changed, the consumer will kill itself and the cronjob will spawn a new one on its next run and only then will handle that message.
+
+Magento 2.3.2 was the first Open Source version coming with some default features making actual use of the message queue system. These are tasks which are triggered using the backend of Magento which can potentially take a while to execute. In order to prevent the webserver or php-fpm to run against a timeout, it was chosen to send these tasks to the message queue system and let these tasks get executed asynchronously. Some examples are:
+
+- Generating coupon codes
+- Mass editing products
+- Exporting data
+- ...
+
+Currently in Magento 2.3.2, 4 consumer processes get spawned by the cron system. Each of these processes take memory and cpu and regularly queries the database (or RabbitMQ if that broker is being used) to see if new messages are available and then process them.
+
+
+## Problem 1: too little control
+
+### The problem
+
+There is currently too little control over these consumer processes.  
+What if people only very irregularly use one of these features mentioned above, let's say they only once a year export all their products to check their inventory levels.  
+Then you have some consumer process sitting there, doing nothing for 364 days in the year, wasting precious cpu cycles and taking up precious memory, until finally once a year the shopowner decides to execute a certain task it can execute.
+
+There is also the potential problem that the consumer process will take up more and more memory due to some memory leak in the code it executes.
```

Comment:
Thats should not be a case, cause if there a memory leak then its an architectural/core bug that have to be fixed. But I understand your point

Reply:
Removed this section.